src/nbl/video/utilities/CComputeBlit.cpp

```diff
@@ -0,0 +1,152 @@
+}
+
+core::smart_refctd_ptr<video::IGPUSpecializedShader> CComputeBlit::createBlitSpecializedShader(const asset::E_FORMAT inImageFormat, const asset::E_FORMAT outImageFormat,
+	const asset::E_FORMAT outImageViewFormat, const asset::IImage::E_TYPE inImageType, const core::vectorSIMDu32& inExtent, const core::vectorSIMDu32& outExtent,
+	const asset::IBlitUtilities::E_ALPHA_SEMANTIC alphaSemantic, const uint32_t workgroupSize, const uint32_t alphaBinCount)
+{
+	char shaderStart[] = "#version 460 core\n#include \"nbl/builtin/glsl/blit/default_compute_blit.comp\"\n";
+	auto shader = core::make_smart_refctd_ptr<asset::CCustomAllocatorCPUBuffer<core::null_allocator<uint8_t>>>(strlen(shaderStart), shaderStart, core::adopt_memory);
+
+	auto cpuShader = core::make_smart_refctd_ptr<asset::ICPUShader>(std::move(shader), asset::ICPUShader::buffer_contains_glsl, asset::IShader::ESS_COMPUTE, "CComputeBlit::createBlitSpecializedShader");
+	
+	const core::vectorSIMDu32 windowDim = getWindowDim(inExtent, outExtent);
+
+	// Fail if any dimension of the window is bigger than workgroup size, forcing us to reuse inovcation to process just a single dimension of that window
+	if ((windowDim.x > workgroupSize) || (windowDim.y > workgroupSize) || (windowDim.z > workgroupSize))
+		return nullptr;
+
+	const uint32_t windowPixelCount = windowDim.x * windowDim.y * windowDim.z;
+	// Fail if I would need to reuse invocations just to process a single window
+	if (windowPixelCount > workgroupSize)
+		return nullptr;
+
+	const uint32_t outChannelCount = asset::getFormatChannelCount(outImageFormat);
+	// It is important not to use asset::getTexelOrBlockBytesize here.
+	// Even though input pixels are stored in the shared memory we use outChannelCount here because only outChannelCount channels of
+	// the input image actually need blitting and hence only those channels are stored in shared memory.
+	const size_t windowSize = static_cast<size_t>(windowPixelCount) * outChannelCount * sizeof(float);
+	// Fail if the window cannot be preloaded into shared memory
+	// Todo(achal): Take smem size as a param of this function?
+	if (windowSize > sharedMemorySize)
+		return nullptr;
+
+	// Todo(achal): All this belongs in validation
+	{
+		const uint32_t inChannelCount = asset::getFormatChannelCount(inImageFormat);
+		assert(outChannelCount <= inChannelCount);
+
+		// inFormat should support SAMPLED_BIT format feature
+	}
+
+	const char* outImageFormatGLSLString = getGLSLFormatStringFromFormat(outImageFormat);
+	const char* glslFormatQualifier = getGLSLFormatStringFromFormat(outImageViewFormat);
+
+	char formatInclude[1024] = "";
+	if (outImageFormat != outImageViewFormat)
+		snprintf(formatInclude, sizeof(formatInclude), "#include <nbl/builtin/glsl/blit/blit/formats/%s.glsl>\n", outImageFormatGLSLString);
+
+	// Todo(achal): Get it from outside
+	const uint32_t wgSize = DefaultBlitWorkgroupSize;
+
+	const uint32_t smemFloatCount = sharedMemorySize / (sizeof(float) * outChannelCount);
```

Comment:
What does that mean?

The intended usage of `sharedMemorySize` is to tell us how much shared memory `CComputeBlit` should use. Where else do you want me to use this shared memory?

Reply:
> What does that mean?

Always using 16kb (or 32kb on some other GPUs) limits the occupancy of a compute kernel.
https://docs.nvidia.com/cuda/cuda-occupancy-calculator/index.html
https://en.wikipedia.org/wiki/CUDA#Version_features_and_specifications

For example, your default workgroup size is 256, to run on  mobiles and achieve good occupancy on desktop GPUs. Why? because:
- Mobile vendors only support 256 as the max WG size
- NVidia experiments with "Maximum number of resident threads per multiprocessor" making it 1024, 1536, 2048 (these are always HIGHER than GL/Vulkan max workgroup sizes) between generations and 512 is the greatest common divisor of them all
- AMD recommends 256

So on an Nvidia max occupancy (workgroups per SM also known as a CU on AMD) is between 4 and 8

however you won't reach max occupancy if each workgroup uses 32kb) of SMem because total amount of SMem available ranges between 64 and 164kb, meaning you would hit 62.5% or even 16.6% occupancy instead of 100%.

> The intended usage of sharedMemorySize is to tell us how much shared memory CComputeBlit should use. Where else do you want me to use this shared memory?

As the user how much smem a workgroup is "allowed" to use, an size the output window accordingly.

Reply:
Resolved.